### PR TITLE
[SPARK-36553][ML] KMeans avoid compute auxiliary statistics for large K

### PR DESCRIPTION
**Pull request by zhengruifeng**
### What changes were proposed in this pull request?

SPARK-31007 introduce an auxiliary statistics to speed up computation in KMeasn.

However, it needs a array of size `k * (k + 1) / 2`, which may cause overflow or OOM when k is too large.

So we should skip this optimization in this case.

### Why are the changes needed?

avoid overflow or OOM when k is too large (like 50,000)

### Does this PR introduce _any_ user-facing change?
No

### How was this patch tested?
existing testsuites

### CODE REVIEW

**Comment by anders-rydbirk** [comment]
@zhengruifeng Thanks for picking this one up!

**Comment by zhengruifeng** [comment]
cc @srowen

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/DistanceMeasure.scala [hunk 1]
```diff
@@ -117,6 +117,17 @@ private[spark] abstract class DistanceMeasure extends Serializable {
     packedValues
   }
 
+  def findClosest(
```
Comment (srowen) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/DistanceMeasure.scala:120]: Is this overload used?

**Review by srowen — COMMENTED** [review]
I don't love the two code paths but maybe this is the easiest fix. It un-optimizes large k though. Is the idea to un-pack that triangular array not viable?

**Comment by zhengruifeng** [comment]
I think I made it too complex.



according to @anders-rydbirk  your description in the ticket:

```
Possible workaround:

    Roll back to Spark 3.0.0 since a KMeansModel generated with 3.0.0 cannot be loaded in 3.1.1.
    Reduce K. Currently trying with 45000.

```

maybe we just need to chang `k * (k + 1) / 2` to `(k.toLong * (k + 1) / 2).toInt`?
```
scala> val k = 50000
val k: Int = 50000

scala> k * (k + 1) / 2
val res8: Int = -897458648

scala> (k.toLong * (k + 1) / 2).toInt
val res9: Int = 1250025000

scala> val k = 45000
val k: Int = 45000

scala> k * (k + 1) / 2
val res10: Int = 1012522500

scala> (k.toLong * (k + 1) / 2).toInt
val res11: Int = 1012522500
```



> Sorry, I guess I mean make it into an array of arrays, not one big array.

@srowen  yes, using arrays of sizes (1, 2, ..., k) is another choice

**Comment by srowen** [comment]
Array sizes can't be long so if it doesn't fit in an int it won't work

**Comment by zhengruifeng** [comment]
there are two limits:

1, array size, must be less than `Int.MaxValue`;

2, its size should fit in memory for initialization and broadcasting.

with --driver-memory=8G, I can not create an array of 1250025000 doubles. If we switch to arrays of arrays, I am afraid it's prone to OOM for large K.

**Comment by zhengruifeng** [comment]
@srowen 

I can switch to array[array[double] if you perfer it, I am netural on it.

my main concern is, this optional statistics may be too large. In this case, k=50,000, it is much larger than the clustering centers (dim=3).

**Comment by srowen** [comment]
Your current design is fine, I trust your judgment

**Review by huaxingao — APPROVED** [review]

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/DistanceMeasure.scala [hunk 2]
```diff
@@ -117,6 +117,17 @@ private[spark] abstract class DistanceMeasure extends Serializable {
...
+      centers: Array[VectorWithNorm],
+      statistics: Option[Array[Double]],
+      point: VectorWithNorm): (Int, Double) = {
+    if (statistics.nonEmpty) {
+      findClosest(centers, statistics.get, point)
+    } else {
+      findClosest(centers, point)
+    }
+  }
```
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/DistanceMeasure.scala:129]: Shall we add the function description like the other existing `def findClosest` functions?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala [hunk 3]
```diff
@@ -52,7 +52,14 @@ class KMeansModel (@Since("1.0.0") val clusterCenters: Array[Vector],
   @transient private lazy val statistics = if (clusterCenters == null) {
     null
   } else {
-    distanceMeasureInstance.computeStatistics(clusterCentersWithNorm)
+    val k = clusterCenters.length
+    val numFeatures = clusterCenters.head.size
+    if (DistanceMeasure.shouldComputeStatistics(k) &&
+      DistanceMeasure.shouldComputeStatisticsLocally(k, numFeatures)) {
```
Comment (dongjoon-hyun) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeansModel.scala:58]: indentation?

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/DistanceMeasure.scala (hunk 2 above)
Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/DistanceMeasure.scala:129]: good point. I will update this PR

**Comment by zhengruifeng** [comment]
I think this should also be back-ported to 3.1/3.2

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/DistanceMeasure.scala (hunk 1 above)
Comment (srowen) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/DistanceMeasure.scala:127]: OK this is a new method but I don't see it called, maybe I'm missing something

**Review by srowen — COMMENTED** [review]
It's adding complexity but I think for a reasonable reason, to fix a perf regression

**Comment by zhengruifeng** [comment]
image

@srowen It is used in both training (in the .ml side) and prediction (in the .mllib side), the switch is done by just changing the type of `stats` in `distanceMeasureInstance.findClosest(centers, stats, point)` from `Array[Double]` to `Option[Array[Double]]`

**Comment by srowen** [comment]
Do existing call sites bind to the new method? I can't see how a new method is called when nothing new calls it, but if you understand it and it works, nevermind

**Comment by zhengruifeng** [comment]
> Do existing call sites bind to the new method?

NO.

existing two methods are used in `DistanceMeasure` and `DistanceMeasureSuite`;

but `def findClosest(centers: Array[VectorWithNorm], point: VectorWithNorm)` is also used in KMeans initialization algorithm `initKMeansParallel` and  `BisectingKMeans`.

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala [hunk 4]
```diff
@@ -269,15 +269,22 @@ class KMeans private (
...
-    val shouldDistributed = centers.length * centers.length * numFeatures.toLong > 1000000L
+    val shouldComputeStats =
+      DistanceMeasure.shouldComputeStatistics(centers.length)
+    val shouldComputeStatsLocally =
+      DistanceMeasure.shouldComputeStatisticsLocally(centers.length, numFeatures)
 
     // Execute iterations of Lloyd's algorithm until converged
     while (iteration < maxIterations && !converged) {
       val bcCenters = sc.broadcast(centers)
-      val stats = if (shouldDistributed) {
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala:277]: previous stats is a Array[Double]

File: mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala [hunk 5]
```diff
@@ -269,15 +269,22 @@ class KMeans private (
...
-        distanceMeasureInstance.computeStatisticsDistributedly(sc, bcCenters)
+      val stats = if (shouldComputeStats) {
```
Comment (zhengruifeng) [inline mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala:280]: Now, it is a Option[Array[Double]]

Comment (zhengruifeng) [reply mllib/src/main/scala/org/apache/spark/mllib/clustering/KMeans.scala:280]: So the following `val (bestCenter, cost) = distanceMeasureInstance.findClosest(centers, stats, point)` will call this new method, without code change in the call sites.

**Comment by huaxingao** [comment]
Merged to master/3.2/3.1. Thanks!

**Comment by zhengruifeng** [comment]
@huaxingao @srowen @dongjoon-hyun  Thanks for reviewing!

**Pull request closed without merging**